### PR TITLE
i18n: treat Infinity and NaN as numeric values

### PR DESCRIPTION
**Pull request by brendankenny**
fixes the crasher part of #9684

The issue was that we were `JSON.parse(JSON.stringify(values))`ing to make sure we weren't modifying the original ICU replacement values, but both `Infinity` and `NaN` get converted to `null` when stringified, which then fails the check that we're only replacing numeric placeholders with numbers.

Instead, this copies over the values one by one to a new object, formatting if needed. Also smushes `_preformatValues` and `_processParsedElements` together because the first was mostly just calling the second, but the first had a way better name.

### CODE REVIEW

File: lighthouse-core/test/lib/i18n/i18n-test.js [hunk 1]
```diff
@@ -265,5 +283,13 @@ describe('i18n', () => {
         // eslint-disable-next-line max-len
         .toThrow(`Provided value "sirNotAppearingInThisString" does not match any placeholder in ICU message "Hello {timeInMs, number, seconds} World"`);
     });
+
+    it('formats correctly with NaN and Infinity numeric values', () => {
+      const helloInfinityStr = str_(UIStrings.helloBytesWorld, {in: Infinity});
+      expect(helloInfinityStr).toBeDisplayString('Hello ∞ World');
```
Comment (brendankenny) [inline lighthouse-core/test/lib/i18n/i18n-test.js]: > `expect(helloInfinityStr).toBeDisplayString('Hello ∞ World');`

I guess this is how it's supposed to get formatted? 🤷‍♀

Comment (exterkamp) [reply lighthouse-core/test/lib/i18n/i18n-test.js]: ◊◊◊

File: lighthouse-core/lib/i18n/i18n.js [hunk 2]
```diff
@@ -217,28 +209,34 @@ function _processParsedElements(icuMessage, argumentElements, values = {}) {
...
     } else if (format.style === 'seconds' && id === 'timeInMs') {
       // Convert all seconds to the correct unit (currently only for `timeInMs`).
-      values[id] = Math.round(value / 100) / 10;
+      formattedValues[id] = Math.round(value / 100) / 10;
     } else if (format.style === 'bytes') {
       // Replace all the bytes with KB.
-      values[id] = value / 1024;
+      formattedValues[id] = value / 1024;
+    } else {
+      // All other numbers passed through unchanged.
```
Comment (exterkamp) [inline lighthouse-core/lib/i18n/i18n.js]:
```suggestion
      // All other numbers (e.g. ∞) passed through unchanged.
```

File: lighthouse-core/lib/i18n/i18n.js [hunk 3]
```diff
@@ -217,28 +209,34 @@ function _processParsedElements(icuMessage, argumentElements, values = {}) {
...
+      formattedValues[id] = value;
     }
   }
 
   // Throw an error if a value is provided but has no placeholder in the message.
   for (const valueId of Object.keys(values)) {
-    // errorCode is a special case always allowed to help ease of LHError use.
-    if (valueId === 'errorCode') continue;
+    if (valueId in formattedValues) continue;
```
Comment (exterkamp) [inline lighthouse-core/lib/i18n/i18n.js]: Nice elimination of a `find`

File: lighthouse-core/lib/i18n/i18n.js (hunk 2 above)
Comment (brendankenny) [reply lighthouse-core/lib/i18n/i18n.js]: Infinity could be passed to the other styles as well (the issue in #9684 is passing it through `bytes`, it's just that `Infinity / 1024 === Infinity` :)

File: lighthouse-core/lib/i18n/i18n.js [hunk 4]
```diff
@@ -188,27 +173,34 @@ function collectAllCustomElementsFromICU(icuElements, seenElementsById = new Map
...
+ * @param {Readonly<Record<string, string | number>>} values
  * @return {Record<string, string | number>}
  */
-function _processParsedElements(icuMessage, argumentElements, values = {}) {
+function _preformatValues(icuMessage, messageFormatter, values) {
+  const elementMap = collectAllCustomElementsFromICU(messageFormatter.getAst().elements);
+  const argumentElements = [...elementMap.values()];
+
+  /** @type {Record<string, string | number>} */
+  const formattedValues = {};
```
Comment (paulirish) [inline lighthouse-core/lib/i18n/i18n.js]: can ya add a comment on why we create a new obj to hold these

File: lighthouse-core/test/lib/i18n/i18n-test.js (hunk 1 above)
Comment (paulirish) [reply lighthouse-core/test/lib/i18n/i18n-test.js]: ∞∞∞∞∞∞∞v∞

File: lighthouse-core/lib/i18n/i18n.js (hunk 4 above)
Comment (brendankenny) [reply lighthouse-core/lib/i18n/i18n.js]: > can ya add a comment on why we create a new obj to hold these

done

File: lighthouse-core/lib/i18n/i18n.js (hunk 2 above)
Comment (brendankenny) [reply lighthouse-core/lib/i18n/i18n.js]: tried to make this comment clearer as well

Comment (brendankenny) [reply lighthouse-core/lib/i18n/i18n.js]: tried to make this comment clearer as well

**Pull request merged**